ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs

```diff
@@ -190,10 +190,10 @@ defaultCodecs ccfg version networkVersion = Codecs {
           (networkVersion >= NodeToClientV_8)
```

Comment:
Perhaps `codecLocalStateQuery` takes a boolean instead of the version for the same reasons you had mentioned about the awkwardness of importing the right types into the encoder's definition?

Reply:
I was suggesting you could change your encoder and decoder to take the `isTopLevelQueryIntroduced` boolean argument, comparable to the `canAcquireTip` flag that `codecLocalStateQuery` already takes.

Maybe that "cuts the knot" with respect to the types; avoiding the awkwardness.

Reply:
I'll admit I haven't studied the details of the awkwardness you mentioned. I'll do that now.

Reply:
I've been thinking a bit about this. It seems key that there are 2 separate layers of code here: the network layer and the consensus layer. `NodeToClientVersion` corresponds to the network layer version. `BlockNodeToClientVersion blk` corresponds to the consensus layer version.

What I've implicitly done in this PR is assume that the addition of a top level query, `Query blk`, is a change to the network layer. Hence I've added a new `NodeToClientVersion` (version 9). If that's correct, then I can continue on with this. I'd suggest just passing the `NodeToClientVersion` to `codecLocalStateQuery` and doing the version check for both `Query blk` and for  `canAcquireTip` based on that.

But is this assumption really correct? Perhaps the addition of `Query blk` is actually a consensus layer change. Indeed the network layer has always been polymorphic over the query type. If the change is part of the network layer, then that would mean I should add a new version to *all* `BlockNodeToClientVersion blk`s instead of just to `NodeToClientVersion`. Then I'd need to touch all thoughs types again every time I add a new constructor to `Query blk`.

Any thoughts on this would be helpful.

Reply:
- See my comment on `queryEncodeResult`; we don't _need_ to worry about serialization at all in this PR. It's up to you whether we do or not, though I would prefer to defer it for this PR so we can work out the serialization details on a narrower PR.
- The semantics of `NodeToClientVersion` and `BlockNodeToClientVersion` accumulate in the opposite direction of the dependency: the `NodeToClientVersion` characterizes the behavior of the whole node (not just the network layer), whereas `BlockNodeToClientVersion` characterizes the behavior of a part of the Consensus part of the node. So if the Consensus part changes, then that means the whole node's behavior changes. (Somewhat tangential, but potentially clarifying remark: the network layer is responsible for negotiating the `NodeToClientVersion` with the peer).
- I agree that "the addition of Query blk is actually a consensus layer change", but because of the above point, that still means we'll need to bump `NodeToClientVersion`. We have to whenever we alter the set of possible queries and responses. Here's a historical example of that:

https://github.com/input-output-hk/ouroboros-network/blob/7c8fac89c8ce4c4f9e5d891c7fc1be6613af3750/ouroboros-network/src/Ouroboros/Network/NodeToClient/Version.hs#L32-L33

- It helps me to think of `NodeToClientVersion` (and its counterpart `NodeToNodeVersion`) as merely the monotonically-increasing natural number that every pair of peers negotiates when connecting. That's its fundamental role and why it needs to reflect every breaking change.
- The evolution of `BlockNodeToClientVersion` versions then is in some sense just a subsequence (not substring) of the evolution of `NodeToClientVersion` versions: its counter corresponds to when the negotiated thing had to change in ways that (only?) `blk`-sensitive Consensus code needs to consider (eg for backwards-compatibility). (That's the number, like 6 in `CardanoNodeToClientVersion6` -- but the RHS of `CardanoNodeToClientVersion6` breaks it down even farther regarding which parts of Consensus are affected).